dianatanase/assignment3/Q1.java

```diff
@@ -0,0 +1,158 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+
+// Create a vertex class to represent the entries in the grid.
+// Each vertex has the associated character and the position in the grid (line and column).
+// For each vertex, store the vertices that can be reached from it (children), going in the 8 directions.
+class Vertex{
+	
+	char ch;
+	int line;
+	int column;
+	ArrayList<Vertex> children = new ArrayList<Vertex>();
+	
+	Vertex(int line, int column, char ch){
+		this.line = line;
+		this.column = column;
+		this.ch = ch;
+	}
+	
+	public void addChild(Vertex v){
+		children.add(v);
+	}
+	
+	public int getLine(){
+		return line;
+	}
+	
+	public int getColumn(){
+		return column;
+	}
+	
+	public ArrayList<Vertex> getChildren(){
+		return children;
+	}
+	
+	public char getChar(){
+		return ch;
+	}	
+}
+
+// A graph represented by its vertices.
+// Vertices are stored in a matrix such that
+// element at (i,j) is the corresponding Vertex of the cell at (i,j) in the given grid.
+// Grid is an n x m matrix of chars.
+class Graph{
+	Vertex[][] vertices;
+	int lines;
+	int columns;
+	
+	// lines and columns are the size of the initial grid used to construct the representation.
+	Graph(int lines, int columns){
+		this.lines = lines;
+		this.columns = columns;
+		vertices = new Vertex[lines][columns];
+	}
+	
+	// Create the vertices and the relation between them.
+	// Return a matrix of Vertex objects.
+	public Vertex[][] createVertices(char[][] grid){
+		
+		// Create a vertex for each cell in the grid.
+		for(int i = 0; i < lines; i++){
+			for(int j = 0; j < columns; j++){
+				Vertex v = new Vertex(i, j, grid[i][j]);
+				vertices[i][j] = v;
+			}
+		}
+		
+		// For each vertex, add its children (in the 8 directions).
+		for(int i = 0; i < lines; i++){
+			for(int j = 0; j < columns; j++){
+				addChildrenToVertex(i, j);	
+			}
+		}
+		
+		return vertices;
+	}
+	
+	// Given position (i,j) of a vertex in vertices matrix, add its children (in the 8 directions).s
+	public void addChildrenToVertex(int i, int j){
+		
+		if(i != 0){
```

Comment:
You can simplify this code by using a couple of for loops to iterate over the 3x3 square and ignore the middle value. Any time you end up with a big list of ifs, the code could probably be simplified.